### PR TITLE
feat: 🎸 @W-8712470@ add env var for target registry

### PR DESCRIPTION
**Pull request by peternhale**
Added a env var, NPM_REGISTRY, that allows users to select which npm
registry is to be used for install and publish actions.

### What does this PR do?
Gives the user the ability to direct certain actions to a specific npm registry by setting env var NPM_REGISTRY.

### CODE REVIEW

File: src/repository.ts [hunk 1]
```diff
@@ -186,6 +189,13 @@ abstract class Repository extends AsyncOptionalCreatable {
     return found;
   }
 
+  private getRegistryForAuth(): string {
+    const registry: string = this.env.getString('NPM_REGISTRY');
```
Comment (mdonnalley) [inline src/repository.ts:193]: you don't need to specify `string` here, it should be implied from `getString`

**Review by mdonnalley — APPROVED** [review]

File: src/repository.ts (hunk 1 above)
Comment (peternhale) [reply src/repository.ts:193]: Will remove

**Review by WillieRuemmele — APPROVED** [review]

**Review by shetzel — APPROVED** [review]

**Review by amphro — APPROVED** [review]

**Comment by peternhale** [comment]
@amphro @mdonnalley I did some additional work when I found out the the original changes just didn't work.

Created an abstraction of a registry to help isolate decisions made around that subject.

File: src/package.ts [hunk 2]
```diff
@@ -58,7 +61,8 @@ export class Package extends AsyncOptionalCreatable {
   }
 
   public retrieveNpmPackage(): NpmPackage {
-    const result = exec(`npm view ${this.name} --json`, { silent: true });
+    const registry = new Registry(this.env.getString('NPM_REGISTRY', this.env.getString('NPM_TOKEN')));
```
Comment (mdonnalley) [inline src/package.ts:64]: Is it necessary to pass in the env vars? It looks like the Registry class is already checking it in the init method

File: src/registry.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/* eslint-disable no-underscore-dangle */
```
Comment (mdonnalley) [inline src/registry.ts:8]: Ideally, I'd prefer to not disable any of the eslint rules

File: src/registry.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+  /**
+   * Compare this.registryUrl against npm configs and write registry entry "registry" entry to <packageDir>/.npmrc
+   * if either is not equal
+   *
+   * @param packageDirectory
+   */
+  public async setNpmRegistry(packageDirectory: string): Promise<void> {
+    if (this.registryEntryLocal !== this.registryUrl || this.registryEntryGlobal !== this.registryUrl) {
+      let npmrc: string[] = await this.readNpmrc(packageDirectory);
```
Comment (mdonnalley) [inline src/registry.ts:57]: `string[]` here is redundant since it's implied from `readNpmrc`

**Review by mdonnalley — CHANGES_REQUESTED** [review]

File: src/package.ts (hunk 2 above)
Comment (peternhale) [reply src/package.ts:64]: 👍 Should be able to use new `new Registry()`, but will verify first.

File: src/registry.ts (hunk 3 above)
Comment (peternhale) [reply src/registry.ts:8]: 👍

File: src/registry.ts (hunk 4 above)
Comment (peternhale) [reply src/registry.ts:57]: 👍

**Review by amphro — APPROVED** [review]

**Review by mdonnalley — APPROVED** [review]

**Comment by amphro** [comment]
Testing notes: I published a couple of different packages to a local verdaccio including some error conditions. Things look good. The `release-package` is waiting for a status.... but everything else is successful... I'll merge in the morning even if it hasn't run yet.

**Pull request merged**